### PR TITLE
feat(`x/marketmap`): refactor all market map consumer code

### PR DESCRIPTION
**Pull request by aljo242**

### CODE REVIEW

File: cmd/slinky/main.go [hunk 1]
```diff
@@ -99,27 +99,31 @@ func runOracle() error {
...
 	var marketCfg mmtypes.MarketMap
 	if marketCfgPath != "" {
 		marketCfg, err = types.ReadMarketConfigFromFile(marketCfgPath)
 		if err != nil {
-			return fmt.Errorf("failed to read market config file: %s", err.Error())
+			return fmt.Errorf("failed to read market config file: %w", err)
+		}
+
+		if err := marketCfg.ValidateBasic(); err != nil {
+			return fmt.Errorf("market config is invalid: %w", err)
```
Comment (davidterpay) [inline cmd/slinky/main.go]: does it make more sense to move this validation logic to `ReadMarketConfigFromFile`? Or does it already do this?

Comment (davidterpay) [reply cmd/slinky/main.go]: We should have the default behavior of `ReadMarketConfigFromFile` validate that the read in config is correct.

File: config/dydx/oracle.json [hunk 2]
```diff
@@ -401,7 +401,7 @@
         "reconnectTimeout": 2000000000,
         "maxQueries": 1,
         "atomic": true,
-        "url": "localhost:1317",
+        "url": "https://dydx-api.polkachu.com",
```
Comment (davidterpay) [inline config/dydx/oracle.json]: remove polkachu

Comment (aljo242) [reply config/dydx/oracle.json]: https://github.com/skip-mev/slinky/pull/294/commits/c6ddff5d5b9c524a1a0c420027e3592fe7f4a16f

File: cmd/slinky/main.go (hunk 1 above)
Comment (aljo242) [reply cmd/slinky/main.go]: https://github.com/skip-mev/slinky/pull/294/commits/b90c76e88d57a6a51b815354223d3023e9997c8c

File: cmd/slinky-config/main.go [hunk 3]
```diff
@@ -84,8 +84,8 @@ var (
 		// -----------------------------------------------------------	//
 		// ---------------------Start API Providers--------------------	//
 		// -----------------------------------------------------------	//
-		binance.Name:       binance.DefaultNonUSMarketConfig,
-		coinbaseapi.Name:   coinbaseapi.DefaultMarketConfig,
+		binance.Name:       binance.DefaultNonUSProviderConfig,
+		coinbaseapi.Name:   coinbaseapi.DefaultProviderConfig,
```
Comment (davidterpay) [inline cmd/slinky-config/main.go:88]: Why is this renamed to `DefaultProviderConfig` instead of `DefaultMarketConfig`?

File: oracle/orchestrator/market_mapper.go [hunk 4]
```diff
@@ -55,6 +55,7 @@ func (o *ProviderOrchestrator) listenForMarketMapUpdates(ctx context.Context) {
 			}
 
 			// Write the market map to the configured path.
+			o.logger.Info("writing updated market map", zap.String("file", o.writeTo))
```
Comment (davidterpay) [inline oracle/orchestrator/market_mapper.go]: can we move this into the `WriteMarketMap` function since the o`writeTo` might be empty?

File: x/marketmap/types/market_test.go [hunk 5]
```diff
@@ -383,9 +383,9 @@ func TestValidateMarketMap(t *testing.T) {
 			name: "second operation is not an index price provider",
 			cfg: types.MarketMap{
 				Tickers: map[string]types.Ticker{
-					constants.BITCOIN_USDT.String(): constants.BITCOIN_USDT,
-					constants.BITCOIN_USD.String():  constants.BITCOIN_USD,
-					constants.USDT_USD.String():     constants.USDT_USD,
+					constants.BITCOIN_USDT.String(): types.Ticker(constants.BITCOIN_USDT),
```
Comment (davidterpay) [inline x/marketmap/types/market_test.go:386]: Why do we need type casting in these?

File: oracle/types/market.go [hunk 6]
```diff
@@ -26,7 +26,7 @@ type (
 		TickerConfigs TickerToProviderConfig
 
 		// OffChainMap is a map of tickers to their respective on-chain markets.
-		OffChainMap map[string]mmtypes.Ticker
+		OffChainMap map[string][]mmtypes.Ticker
```
Comment (davidterpay) [inline oracle/types/market.go:29]: Why does the `OffChainMap` now correspond to a list of tickers?

File: oracle/types/market.go [hunk 7]
```diff
@@ -86,14 +82,21 @@ func NewProviderMarketMap(name string, tickerConfigs TickerToProviderConfig) (Pr
 			return ProviderMarketMap{}, fmt.Errorf("expected provider config name %s, got %s", name, config.Name)
 		}
 
-		offChainMap[config.OffChainTicker] = ticker
+		if offChainMap[config.OffChainTicker] == nil {
+			offChainMap[config.OffChainTicker] = []mmtypes.Ticker{ticker}
+			continue
+		}
+
+		offChainMap[config.OffChainTicker] = append(offChainMap[config.OffChainTicker], ticker)
```
Comment (davidterpay) [inline oracle/types/market.go:90]: This doesn't right, the mapping should be 1:1 with the tickerConfigs

File: oracle/types/market.go [hunk 8]
```diff
@@ -110,8 +113,8 @@ func (pmm *ProviderMarketMap) ValidateBasic() error {
 	if len(pmm.Name) == 0 {
 		return fmt.Errorf("provider name cannot be empty")
 	}
-	if len(pmm.OffChainMap) != len(pmm.TickerConfigs) {
-		return fmt.Errorf("off-chain map length mismatch")
+	if len(pmm.OffChainMap) > len(pmm.TickerConfigs) {
+		return fmt.Errorf("off-chain map length invalid %d>%d, %v %v", len(pmm.OffChainMap), len(pmm.TickerConfigs), pmm.OffChainMap, pmm.TickerConfigs)
```
Comment (davidterpay) [inline oracle/types/market.go:117]: Why would the lengths be different?

File: pkg/math/oracle/median.go [hunk 9]
```diff
@@ -121,56 +111,45 @@ func (m *MedianAggregator) AggregateData() {
...
-			m.logger.Error(
-				"no operations",
-				zap.String("ticker", target.String()),
-			)
-
-			continue
-		}
-
+	convertedPrices := make([]*big.Int, 0, len(market.ProviderConfigs))
+	for _, config := range market.ProviderConfigs {
```
Comment (davidterpay) [inline pkg/math/oracle/median.go:127]: This pre-alloc on the size won't affect any price calculations right?

File: pkg/math/oracle/utils.go [hunk 10]
```diff
@@ -4,54 +4,72 @@ import (
...
-	price, ok := cache[ticker]
-	if !ok {
-		return nil, fmt.Errorf("missing %s price for ticker: %s", operation.Provider, ticker.String())
+	return scaledPrice, nil
+}
+
+// GetIndexPrice returns the aggregated index price.
+func (m *MedianAggregator) GetIndexPrice(
+	providerConfig mmtypes.ProviderConfig,
+) (*big.Int, error) {
```
Comment (davidterpay) [inline pkg/math/oracle/utils.go]: small nit: I think this should just accept the `NormalizebyPair`

Comment (davidterpay) [reply pkg/math/oracle/utils.go]: We already know that this is only called if normalize by pair is not nil so you can just pass in the de-ref pair here.

Comment (davidterpay) [reply pkg/math/oracle/utils.go]: simplifies the API

File: pkg/math/oracle/utils.go [hunk 11]
```diff
@@ -4,54 +4,72 @@ import (
...
+	if !ok {
+		return nil, fmt.Errorf("missing index price for ticker: %s", targetTicker.String())
+	}
+
+	scaledPrice, err := ScaleUpCurrencyPairPrice(targetTicker.Decimals, price)
+	if err != nil {
+		return nil, err
 	}
 
 	return scaledPrice, nil
```
Comment (davidterpay) [inline pkg/math/oracle/utils.go]:
```suggestion
	return ScaleUpCurrencyPairPrice(targetTicker.Decimals, price)
```

File: pkg/math/oracle/median.go (hunk 9 above)
Comment (aljo242) [reply pkg/math/oracle/median.go:127]: yes, this is just the "capacity" not the length

slice initailization is `make(type, len, capacity)`

File: oracle/types/market.go (hunk 8 above)
Comment (aljo242) [reply oracle/types/market.go:117]: The wonky behavior of duplicate feeds (this goes away with our proposed fix

File: oracle/types/market.go (hunk 7 above)
Comment (aljo242) [reply oracle/types/market.go:90]: The wonky behavior of duplicate feeds (this goes away with our proposed fix

File: cmd/slinky-config/main.go (hunk 3 above)
Comment (aljo242) [reply cmd/slinky-config/main.go:88]: They are all going to be renamed provider config - but this must have accidentally been included

https://github.com/skip-mev/slinky/pull/296

File: x/marketmap/types/market_test.go (hunk 5 above)
Comment (aljo242) [reply x/marketmap/types/market_test.go:386]: in this temp state, we have two types of Tickers (sadge)

marketmap.Ticker and mm2.Ticker

(this goes away once we move mm2 -> marketmap

**Review by davidterpay — CHANGES_REQUESTED** [review]
Think we should close this since the refactoring im working on directly conflicts

**Comment by aljo242** [comment]
> Think we should close this since the refactoring im working on directly conflicts

i'll leave open - but as a draft
